### PR TITLE
fix _floatToI32/_doubleToI64 in FP emulation

### LINKED ISSUE #7497: FPHelper float-int reinterpretation polyfill bug
the conversion near power of two is mostly wrong.
my fix and test code:
https://github.com/HaxeFoundation/haxe/pull/6697#issuecomment-416920617

### PR DESCRIPTION
**Pull request by R32**
close #7497

here is the test:
```haxe
package;

import haxe.io.FPHelper;
import Math.isNaN;

class Main {

    static function pot(n) return Math.pow(2, n);
    static function hexI32(b: Int) return StringTools.hex(b, 8);

    static function i2f(i) return @:privateAccess FPHelper._i32ToFloat(i);
    static function f2i(f) return @:privateAccess FPHelper._floatToI32(f);

    static function d2ii(d) return @:privateAccess FPHelper._doubleToI64(d);
    static function ii2d(lo, hi) return @:privateAccess FPHelper._i64ToDouble(lo, hi);

    static function rtsingle(i:Int, n:Float, name:String) {
        var ai = f2i(n);
        var af = i2f(ai);
        var bi = FPHelper.floatToI32(af);
        var ci = FPHelper.floatToI32(n);
        if((n != af && !(isNaN(n) && isNaN(af))) || ai != bi || ai != ci)
            throw('"$name" i:$i, n:$n, af:$af');
    }

    static function rtdouble(i:Int, n:Float, name:String){
        var ai = d2ii(n);
        var ad = ii2d(ai.low, ai.high);
        var bi = FPHelper.doubleToI64(ad);
        var ci = FPHelper.doubleToI64(n);
        if((n != ad && !(isNaN(n) && isNaN(ad))) || ai != bi || ai != ci)
            throw('"$name" i:$i, n:$n, ad:$ad');
    }

    static function main() {
        var NA = -9999;
        // float32 to int32
        rtsingle(NA, 0, "s 0");
        rtsingle(NA, -0.0, "s -0");
        rtsingle(NA, pot(-149), "s min");
        rtsingle(NA, -pot(-149), "s -min");
        rtsingle(NA, pot(127) * (2 - pot(-23)), "s max");
        rtsingle(NA, -pot(127) * (2 - pot(-23)), "s -max");
        rtsingle(NA, Math.NEGATIVE_INFINITY, "-infinity");
        rtsingle(NA, Math.POSITIVE_INFINITY, "infinity");
        rtsingle(NA, Math.NaN, "nan");
        for(i in -149...128){ // min denormalized float to max normalized pot float
            var n = pot(i);
            var uulp = if(i < -126) pot(-149) else pot(i - 23);
            var dulp = if(i <= -126) pot(-149) else pot(i - 24);
            rtsingle(i, n, "s pot");
            rtsingle(i, n + uulp, "s pot+");
            rtsingle(i, n - dulp, "s pot-");
            rtsingle(i, n - 2 * dulp, "s pot--");
        }

        // float64 to int64
        rtdouble(NA, 0, "s 0");
        rtdouble(NA, -0.0, "s -0");
        rtdouble(NA, pot(-1075), "d min");
        rtdouble(NA, -pot(-1075), "d -min");
        rtdouble(NA, pot(1023) * (2 - pot( -52)), "d max");
        rtdouble(NA, -pot(1023) * (2 - pot( -52)), "d -max");
        rtdouble(NA, Math.POSITIVE_INFINITY, "d Inf");
        rtdouble(NA, Math.NEGATIVE_INFINITY, "d -Inf");
        rtdouble(NA, Math.NaN, "d NaN");
        for(i in -1075...1024){ // min denormalized double to max normalized pot double
            var n = pot(i);
            var uulp = if(i < -1022) pot(-1075) else pot(i - 52);
            var dulp = if(i <= -1022) pot(-1075) else pot(i - 53);
            rtdouble(i, n, "d pot");
            rtdouble(i, n + 2 * uulp, "d pot+");
            rtdouble(i, n - dulp, "d pot-");
            rtdouble(i, n - 2 * dulp, "d pot--");
        }
    }
}
```

### CODE REVIEW

**Comment by Simn** [comment]
Could you integrate your test into your unit tests in tests/unit?

**Comment by R32** [comment]
It's hard to run unit.test separately.

File: tests/unit/src/unit/issues/Issue7497.hx [hunk 1]
```diff
@@ -0,0 +1,82 @@
+package unit.issues;
+
+import haxe.io.FPHelper;
+import Math.isNaN;
+
+class Issue7497 extends unit.Test {
+#if js
```
Comment (Simn) [inline tests/unit/src/unit/issues/Issue7497.hx:7]: Why is this here?

Comment (R32) [reply tests/unit/src/unit/issues/Issue7497.hx:7]: you mean `#if js` ? since some platforms can't access those methods. and I have only tested it in nodejs/chrome/IE8

Comment (Simn) [reply tests/unit/src/unit/issues/Issue7497.hx:7]: I think FPHelper should work across all targets, and if it doesn't it should be fixed accordingly. Changing it for all targets but then testing it only on JS is not good, even if it's the target that cares about it the most.

Could you remove that `#if js` so we can see if something fails?

Comment (R32) [reply tests/unit/src/unit/issues/Issue7497.hx:7]: the problem is: ` Std.int(bigFloat)` in `_doubleToI64`
https://github.com/HaxeFoundation/haxe/blob/db6823aa93873f475054c33cc9abc2bb3a8b2a83/std/haxe/io/FPHelper.hx#L101-L103

**Comment by Simn** [comment]
This branch needs an update. Also, I still maintain that we have to align the affected targets with regards to changes and tests: If we want to change and test this only for JS that's fine, but we can't make the change for all targets but test only on JS.

**Comment by R32** [comment]
It only passed the test in JS, at least in JS it will no longer throw errors, and I feel it will only be used in old IE browsers.

On other platforms I only tested flash and hl, but unfortunately they all failed when some expr are `±Infinity`(you can't delect a float32 Infinity value by `Math.isFinite`), and I can't find any solution to deal with it. I'm sorry :(

**Comment by R32** [comment]
It seems out of date
